### ISSUE TITLE
Change generated file name for compositions

### ISSUE THREAD

**Issue by warpech** [Starcounter/Blending#340]
Implementing https://github.com/Starcounter/Blending/issues/338 will allow us to improve the default generated naming scheme for files.

Right now we have: https://github.com/Starcounter/Blending/blob/381d38537c2abb9d5cf6c788cbb4b8be16512127/src/BlendingCommon/HtmlViewCompositionHelper.cs#L42-L62

For example, the input:

```
/PriceConfiguration/views/VendibleThumbnailView.html
/Showroom/views/SearchItem.html
/Showroom/views/ProductThumbnail.html
/Search/viewmodels/SearchItem.html
/Products/views/SearchRowPage.html
/PriceConfiguration/views/VendiblePriceView.html
```

Which gives you names like: 

`_PriceConfiguration_views_VendibleThumbnailView+VendiblePriceView+SearchRowPage+SearchItem+ProductThumbnail+SearchItem.html`

Alternative could be to only give path to the main view, and then App Names of attached views, and then an index number if needed, like:

`_PriceConfiguration_views_VendibleThumbnailView + Showroom + Showroom + Search + Products + PriceConfiguration.html`

In case of collision, the next file would be called:

`_PriceConfiguration_views_VendibleThumbnailView + Showroom + Showroom + Search + Products + PriceConfiguration (1).html`

**Comment by warpech** [comment]
Or, since there always will be collisions, maybe choose something that won't grow, such as:

`<URI of main view> + <number> views + (<GUID>).html`

E.g.

`_PriceConfiguration_views_VendibleThumbnailView.html + 5 views (composition 4cf8f70f-6a61-495f-9ee9-93c609e2c0a1).html`

**Comment by warpech** [comment]
Or, since there always will be collisions, maybe choose something that won't grow, such as:

`<URI of main view> + <number> views + (composition <GUID>).html`

E.g.

`_PriceConfiguration_views_VendibleThumbnailView.html + 5 views (composition 4cf8f70f-6a61-495f-9ee9-93c609e2c0a1).html`

@Starcounter/sc-apps, @Starcounter/webplatformteam, @miyconst, @mmnosek any opinion?

**Comment by miyconst** [comment]
>`composition <GUID>`

Will this be the same every time?

**Comment by warpech** [comment]
> Will this be the same every time?

It will be uniquely generate when you first set a custom composition for blended views. Then it will prevail upon edits.

**Comment by miyconst** [comment]
>It will be uniquely generate when you first set a custom composition for blended views. Then it will prevail upon edits.

This means keeping the files on the disk, but re-creating a database will not preserve the compositions. 
Moreover, every new guid value will screw up the git history. It's not the desired behavior.

**Comment by warpech** [comment]
I talked to Konstantin on Slack, and what he proposes is:

- don't have any stored GUID
- the HTML views matching to a composition should not come from the filename nor from a GUID lookup, but from a meta data ("front matter") in the HTML file content, like:

```html
<!-- [ "/People/viewmodels/PersonPage.html", "/App2/ViewModels/PersonTabsPage.html", "/Images/viewmodels/EditableIllustrationsPage.html", "/App3.Skills/ViewModels/SubjectPage.html" ] -->

<link href="/App2/App2.css" rel="stylesheet" />

....
```

I can agree to that. I thought we will go there anyway, but with GUID as an intermediate step. Konstantin convinced me that the intermediate step is not needed.

My current proposal is to call files by the pattern:

`<URI of main view> + <number> views.composition.html`

But ship this change after we store the front matter in the HTML file content.

**Comment by miyconst** [comment]
That sounds about right. I  think we should also involve @malx122 into the discussion.

**Comment by tomalec** [comment]
> My current proposal is to call files by the pattern:
> `<URI of main view> + <number> views.composition.html`

What will be the file name, if you have many compositions for same view and number of apps, but different apps?, like
-`<!-- [ "/People/viewmodels/PersonPage.html", "/App2/ViewModels/PersonTabsPage.html" ] -->`
-`<!-- [ "/People/viewmodels/PersonPage.html", "/App3/ViewModels/AgentPage.html" ] -->`

Is the file name just a convention, and all files will be ready anyway and parsed according to the meta tags?

**Comment by miyconst** [comment]
@tomalec, I thought that here:

>`<URI of main view> + <number> views.composition.html`

- `<URI of main view>` - main view URI / name or whatever, which can be used to uniquely identify a view.
- `<number>` - just a sequence number, like `001`, `002`, `...`, `999`.

When an HTML file is updated, it's main view is identified from the file name, the attached views are read from the file content itself, then the composition is uniquely identified and updated.

Does it make sense?

**Comment by tomalec** [comment]
I think you meant `<URI of main view> + <number>`~views~`.composition.html`.

----

What about 
`<URI of main view> + <number> views.<index>.composition.html`
- `<URI of main view>` - main view URI / name or whatever, which can be used to uniquely identify a view,
- `<number>` - number of sibiling viewsviews,
- `<index>` - just a sequence number, like `001`, `002`, `...`, `999`.

Would then app track  those sequence numbers to be able to delete the file once composition is removed in UI?

**Comment by miyconst** [comment]
>Would then app track those sequence numbers to be able to delete the file once composition is removed in UI?

No, I want the files to be the source of truth, no files should be removed if some compositions are removed from the Blender APP.

**Comment by warpech** [comment]
> No, I want the files to be the source of truth, no files should be removed if some compositions are removed from the Blender APP.

There is no app by that name. If you mean BlendingEditor, that app should work with the source of truth - that means it should modify or delete the files.

The file name doesn't really matter. What matters is the meta part of the file content.

**Comment by miyconst** [comment]
@warpech, @tomalec let's discuss this issue offline when you come to Stockholm.

**Comment by PatrykSzwer** [comment]
1. I also strongly agree with 
>  that app should work with the source of truth - that means it should modify or delete the files.

Removing composition from blending app should also delete a file to avoid having thousands of files after a while and to have control over them.

2. about the name that @tomalec proposed, I understand that this index number will be used in case of 2 compositions with the same mainView but different attached views (although still the same number of them)? If so then as I understand we will be forced to use this metadata inside the files to identify which composition is which and with what attached views. Which is not bad or anything, just pointing that out.  In that case, as @warpech said, name doesn't really matter, only the metadata which blending editor will have to read from each file.
